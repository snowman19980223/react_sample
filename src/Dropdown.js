import React from 'react'
import PropTypes from 'prop-types'
import Details from './Details'
import Button from './Button'
import Caret from './Caret'

const arrowStyles = {
  content: '',
  border: '4px solid',
  borderRightColor: 'transparent',
  borderLeftColor: 'transparent',
  borderBottomColor: 'transparent',
  width: '0',
  height: '0'
}

export default function Dropdown({title, scheme, children}) {
  return (
    <div className="BtnGroup">
      <Details className="details-reset BtnGroup-form d-flex">
        {({toggle}) => (
          <React.Fragment>
            <Button tag="summary" scheme={scheme} grouped onClick={toggle}>
              {title} <div className="d-inline-block v-align-middle" style={arrowStyles} />
            </Button>
            <div
              className="border box-shadow position-absolute px-3 py-2 bg-white mt-1 rounded-1 list-style-none"
              style={{zIndex: 99999}}
            >
              <div className="position-absolute ml-0 mt-n2">
                <Caret edge="top" align="start" />
              </div>
              {children}
            </div>
          </React.Fragment>
        )}
      </Details>
    </div>
  )
}

Dropdown.propTypes = {
  children: PropTypes.node,
  scheme: Button.propTypes.scheme,
  title: PropTypes.string
}
