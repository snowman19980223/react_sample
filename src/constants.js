import * as styledSystem from 'styled-system'
import systemPropTypes from '@styled-system/prop-types'
import themeGet from '@styled-system/theme-get'
import theme from './theme'

const {get: getKey, compose} = styledSystem

export const get = key => themeGet(key, getKey(theme, key))

export const COMMON = compose(
  styledSystem.space,
  styledSystem.color,
  styledSystem.display
)
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.color
}
export const BORDER = compose(
  styledSystem.border,
  styledSystem.shadow
)
BORDER.propTypes = {
  ...systemPropTypes.border,
  ...systemPropTypes.shadow
}

// these are 1:1 with styled-system's API now,
// so you could consider dropping the abstraction
export const TYPOGRAPHY = styledSystem.typography
export const LAYOUT = styledSystem.layout
export const POSITION = styledSystem.position
export const FLEX = styledSystem.flexbox

TYPOGRAPHY.propTypes = systemPropTypes.typography
LAYOUT.propTypes = systemPropTypes.layout
POSITION.propTypes = systemPropTypes.position
FLEX.propTypes = systemPropTypes.flexbox
