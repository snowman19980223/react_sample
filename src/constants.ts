import systemPropTypes from '@styled-system/prop-types'
import {themeGet} from '@styled-system/theme-get'
import PropTypes from 'prop-types'
import React from 'react'
import * as styledSystem from 'styled-system'
import theme from './theme'

interface StyleFn extends styledSystem.styleFn {
  propTypes?: React.WeakValidationMap<any>
}

const {get: getKey, compose, system} = styledSystem

export const get = (key: string) => themeGet(key, getKey(theme, key))

// Common props

export const COMMON: StyleFn = compose(styledSystem.space, styledSystem.color, styledSystem.display)

export interface SystemCommonProps
  extends styledSystem.ColorProps,
    styledSystem.SpaceProps,
    styledSystem.DisplayProps {}

COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.color
}

// Typography props

const whiteSpace = system({
  whiteSpace: {
    property: 'whiteSpace'
    // cssProperty: 'whiteSpace',
  }
})

export const TYPOGRAPHY: StyleFn = compose(styledSystem.typography, whiteSpace)

export interface SystemTypographyProps extends styledSystem.TypographyProps {
  whiteSpace?: 'normal' | 'nowrap' | 'pre' | 'pre-wrap' | 'pre-line'
}

TYPOGRAPHY.propTypes = {
  ...systemPropTypes.typography,
  whiteSpace: PropTypes.oneOf(['normal', 'nowrap', 'pre-wrap', 'pre', 'pre-line'])
}

// Border props

export const BORDER: StyleFn = compose(styledSystem.border, styledSystem.shadow)

export interface SystemBorderProps extends styledSystem.BorderProps, styledSystem.ShadowProps {}

BORDER.propTypes = {
  ...systemPropTypes.border,
  ...systemPropTypes.shadow
}

// Layout props

export const LAYOUT: StyleFn = styledSystem.layout

export type SystemLayoutProps = styledSystem.LayoutProps

LAYOUT.propTypes = systemPropTypes.layout

// Position props

export const POSITION: StyleFn = styledSystem.position

export type SystemPositionProps = styledSystem.PositionProps

POSITION.propTypes = systemPropTypes.position

// Flex props

export const FLEX: StyleFn = styledSystem.flexbox

export type SystemFlexProps = styledSystem.FlexboxProps

FLEX.propTypes = systemPropTypes.flexbox

// Grid props

export const GRID: StyleFn = styledSystem.grid

export type SystemGridProps = styledSystem.GridProps

GRID.propTypes = systemPropTypes.grid
