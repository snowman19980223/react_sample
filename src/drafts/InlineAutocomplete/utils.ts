import getCaretCoordinates from '@koddsson/textarea-caret'
import {Children, EventHandler, SyntheticEvent} from 'react'

import {Coordinates, ShowSuggestionsEvent, Suggestion, TextInputCompatibleChild, Trigger} from './types'

const singleWordTriggerTerminators = new Set([' ', '\n'])
const multiWordTriggerTerminators = new Set(['.', '\n'])

const isWhitespace = (char: string) => /\s/.test(char)

/**
 * Calculate whether or not suggestions should be shown based on the given state of the
 * input. If they should be shown, returns the show event.
 */
export const calculateSuggestionsQuery = (
  triggers: Array<Trigger>,
  text: string,
  caretLocation: number
): ShowSuggestionsEvent | null => {
  // Build backwards from the caret location until the most recent trigger character or terminator
  for (
    let i = caretLocation - 1, query = '', potentialTriggers = triggers;
    i >= 0 && potentialTriggers.length > 0;
    i--
  ) {
    const character = text[i]

    if (singleWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => t.multiWord)
    if (multiWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => !t.multiWord)

    for (const trigger of potentialTriggers.filter(t => character === t.triggerChar)) {
      // Trigger chars must always be preceded by whitespace or be the first character in the input,
      // and even a multi-word query cannot start with whitespace
      if ((i === 0 || isWhitespace(text[i - 1])) && !isWhitespace(query[0])) return {trigger, query}

      potentialTriggers = potentialTriggers.filter(t => t !== trigger)
    }

    query = character + query
  }

  return null
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}

export const getSuggestionValue = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.value

export const getSuggestionKey = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.key ?? suggestion.value

/**
 * Replace a section of a string.
 */
export const replaceSlice = (
  original: string,
  [startInclusive, endExclusive]: [number, number],
  replacement: string
) => {
  const before = original.substring(0, startInclusive)
  const after = original.substring(endExclusive)
  return before + replacement + after
}

/**
 * Attempts to assert that the child element is of a supported type. This can't be enforced
 * by the type system so it has to be done as a runtime check. This isn't foolproof - a
 * component that forwards a ref to a correct element but does not forward event handlers
 * will not work. But it's the best we can reasonably do.
 */
export function requireChildrenToBeInput(
  child: React.ReactElement,
  childRef: React.RefObject<HTMLElement>
): TextInputCompatibleChild {
  Children.only(child) // Assert that the child is lonely
  if (
    // There is no way to know what type the underlying child is until it mounts, so this
    // will always pass on first render before failing on the second render
    childRef.current &&
    !(childRef.current instanceof HTMLInputElement) &&
    !(childRef.current instanceof HTMLTextAreaElement)
  ) {
    throw new TypeError(
      `AutocompleteTextarea child must be a component that forwards a ref and props to an <input> or <textarea> element.`
    )
  }
  return child
}

/**
 * Combine several event handlers into one. The last handler in the list is called first
 * and no further handlers will be called if `event.preventDefault()` is called.
 */
export const augmentHandler =
  <E extends SyntheticEvent>(...handlers: Array<EventHandler<E> | undefined>) =>
  (event: E) => {
    for (const handler of [...handlers].reverse()) {
      if (!event.isDefaultPrevented()) handler?.(event)
    }
  }
