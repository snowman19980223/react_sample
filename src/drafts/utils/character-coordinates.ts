import getCaretCoordinates from '@koddsson/textarea-caret'

export type Coordinates = {
  top: number
  left: number
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}
