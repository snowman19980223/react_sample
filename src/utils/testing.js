import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

enzyme.configure({adapter: new Adapter()})

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * console.error() is ugly af in jest; mock it with a noop,
 * and return the mock so that you can restore it:
 *
 * ```js
 * const mock = silenceConsoleError(jest)
 * // do stuff that calls console.error()
 * mock.mockRestore() // restore console.error()
 * ```
 *
 * Note: we need to pass `jest` as a reference here because
 * jest's mocking functions are bound to a module at runtime, and
 * need to be called as methods rather than pure functions
 * exported from the jest-mock module.
 */
export function silenceConsoleError(jest) {
  return jest.spyOn(console, 'error').mockImplementation(jest.fn())
}

export function mount(component) {
  return enzyme.mount(component)
}
