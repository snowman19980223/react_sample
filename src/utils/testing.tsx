import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const readFile = promisify(require('fs').readFile)

export const COMPONENT_DISPLAY_NAME_REGEX = /^[A-Z][A-Za-z]+(\.[A-Z][A-Za-z]+)*$/

enzyme.configure({adapter: new Adapter()})

export function mount(component: React.ReactElement) {
  return enzyme.mount(component)
}

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toImplementSystemProps: (systemProps: any) => boolean
      toImplementSxProp: () => boolean
      toImplementSxBehavior: () => boolean
      toSetDefaultTheme: () => boolean
      toSetExports: (exports: Record<string, string>) => boolean
    }
  }
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component: React.ReactElement) {
  return renderer.create(component).toJSON() as renderer.ReactTestRendererJSON
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * using react-test-renderer and return the root node
 * ```
 */
export function renderRoot(component: React.ReactElement) {
  return renderer.create(component).root
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component: React.ReactElement): string {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node: React.ReactElement, klass: string): boolean {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node: React.ReactElement, theme = defaultTheme): renderer.ReactTestRendererJSON {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value: number | string): string {
  return typeof value === 'number' ? `${value}px` : value
}

export function percent(value: number | string): string {
  return typeof value === 'number' ? `${value}%` : value
}

export function renderStyles(node: React.ReactElement): any {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className: string): Record<string, string | Record<string, string>> {
  const div = document.createElement('div')
  div.className = className

  const computed = {} as any
  for (const sheet of document.styleSheets) {
    // CSSRulesLists assumes every rule is a CSSRule, not a CSSStyleRule
    for (const rule of sheet.cssRules) {
      if (rule instanceof CSSMediaRule) {
        readMedia(rule)
      } else if (rule instanceof CSSStyleRule) {
        readRule(rule, computed)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function matchesSafe(node: HTMLDivElement, selector: string) {
    if (!selector) {
      return false
    }
    try {
      return node.matches(selector)
    } catch (error) {
      return false
    }
  }

  function readRule(rule: CSSStyleRule, dest: any) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule: CSSMediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    // const dest = computed[key] || (computed[key] = {})
    const dest = {}
    for (const rule of mediaRule.cssRules) {
      if (rule instanceof CSSStyleRule) {
        readRule(rule, dest)
      }
    }

    // Don't add media rule to computed styles
    // if no styles were actually applied
    if (Object.keys(dest).length > 0) {
      computed[key] = dest
    }
  }
}
/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node: React.ReactElement) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node: React.ReactElement) {
  return getProps(node).className
}

export function getClasses(node: React.ReactElement) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export async function loadCSS(path: string) {
  const css = await readFile(require.resolve(path), 'utf8')
  const style = document.createElement('style')
  style.setAttribute('data-path', path)
  style.textContent = css
  document.head.appendChild(style)
  return style
}

export function unloadCSS(path: string) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}

// If a component requires certain props or other conditions in order
// to render without errors, you can pass a `toRender` function that
// returns an element ready to be rendered.

interface Options {
  skipAs?: boolean
  skipSx?: boolean
}

interface BehavesAsComponent {
  Component: React.ComponentType<any>
  systemPropArray: any[]
  toRender?: () => React.ReactElement
  options?: Options
}

export function behavesAsComponent({Component, systemPropArray, toRender, options}: BehavesAsComponent) {
  options = options || {}

  const getElement = () => (toRender ? toRender() : <Component />)

  if (!options.skipSx) {
    it('implements sx prop behavior', () => {
      expect(getElement()).toImplementSxBehavior()
    })
  }

  if (!options.skipAs) {
    it('respects the as prop', () => {
      const As = React.forwardRef<HTMLDivElement>((_props, ref) => <div className="as-component" ref={ref} />)
      const elem = React.cloneElement(getElement(), {as: As})
      expect(render(elem)).toEqual(render(<As />))
    })
  }

  it('sets a valid displayName', () => {
    expect(Component.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
  })

  it('sets the default theme', () => {
    expect(getElement()).toSetDefaultTheme()
  })

  it('renders consistently', () => {
    expect(render(getElement())).toMatchSnapshot()
  })
}

export function checkExports(path: string, exports: Record<any, any>): void {
  it('has declared exports', () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const mod = require(`../${path}`)
    expect(mod).toSetExports(exports)
  })
}
