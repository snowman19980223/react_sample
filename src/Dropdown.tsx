import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Button, {ButtonProps} from './Button'
import Details, {DetailsProps} from './Details'
import useDetails from './hooks/useDetails'
import {COMMON, get, SystemCommonProps} from './constants'
import getDirectionStyles from './DropdownStyles'
import theme from './theme'
import sx, {SxProp} from './sx'
import {ComponentProps} from './utils/types'

const StyledDetails = styled(Details)`
  position: relative;
  display: inline-block;
`

export type DropdownProps = DetailsProps

const Dropdown = ({children, className, ...rest}: DropdownProps) => {
  const {getDetailsProps} = useDetails({closeOnOutsideClick: true})
  return (
    <StyledDetails className={className} {...getDetailsProps()} {...rest}>
      {children}
    </StyledDetails>
  )
}

export type DropdownButtonProps = ButtonProps

const DropdownButton = ({children, ...rest}: DropdownButtonProps) => {
  return (
    <Button as="summary" aria-haspopup="true" {...rest}>
      {children}
      <DropdownCaret />
    </Button>
  )
}

const DropdownCaret = styled.div<SystemCommonProps & SxProp>`
  border: 4px solid transparent;
  margin-left: 12px;
  border-top-color: currentcolor;
  border-bottom-width: 0;
  content: '';
  display: inline-block;
  height: 0;
  vertical-align: middle;
  width: 0;
  ${COMMON};
  ${sx};
`

type StyledDropdownMenuProps = {
  direction?: 'ne' | 'e' | 'se' | 's' | 'sw' | 'w'
} & SystemCommonProps &
  SxProp

const DropdownMenu = styled.ul<StyledDropdownMenuProps>`
  background-clip: padding-box;
  background-color: ${get('colors.white')};
  border: 1px solid rgba(27, 31, 35, 0.15);
  border-radius: ${get('radii.2')};
  box-shadow: 0 3px 12px rgba(27, 31, 35, 0.15);
  left: 0;
  list-style: none;
  margin-top: 2px;
  padding: 5px 0 5px 0 !important; //TODO: fix this override on our markdown styles
  position: absolute;
  top: 100%;
  width: 160px;
  z-index: 100;

  &::before {
    position: absolute;
    display: inline-block;
    content: '';
  }

  &::after {
    position: absolute;
    display: inline-block;
    content: '';
  }

  &::before {
    border: 8px solid transparent;
    border-bottom-color: ${get('colors.blackfade15')};
  }

  &::after {
    border: 7px solid transparent;
    border-bottom-color: ${get('colors.white')};
  }

  // stylelint-disable-next-line selector-max-type
  > ul {
    list-style: none;
  }
  ${props => (props.direction ? getDirectionStyles(props.theme, props.direction) : '')};
  ${COMMON};
  ${sx};
`

const DropdownItem = styled.li`
  display: block;
  padding: ${get('space.1')} 10px ${get('space.1')} 15px;
  overflow: hidden;
  color: ${get('colors.gray.9')};
  text-overflow: ellipsis;
  white-space: nowrap;
  a {
    color: ${get('colors.gray.9')};
    text-decoration: none;
    display: block;
    overflow: hidden;
    color: ${get('colors.gray.9')};
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  &:focus,
  a:focus {
    color: ${get('colors.white')};
    text-decoration: none;
    background-color: ${get('colors.blue.5')};
  }

  &:hover,
  &:hover a {
    color: ${get('colors.white')};
    text-decoration: none;
    background-color: ${get('colors.blue.5')};
    outline: none;
  }
  ${COMMON};
  ${sx};
`

DropdownMenu.propTypes = {
  direction: PropTypes.oneOf(['ne', 'e', 'se', 's', 'sw', 'w']),
  ...COMMON.propTypes,
  ...sx.propTypes
}

DropdownMenu.defaultProps = {
  direction: 'sw',
  theme
}
DropdownMenu.displayName = 'Dropdown.Menu'

DropdownItem.defaultProps = {theme}
DropdownItem.propTypes = {
  ...COMMON.propTypes,
  ...sx.propTypes
}
DropdownItem.displayName = 'Dropdown.Item'

DropdownButton.defaultProps = {theme, ...Button.defaultProps}
DropdownButton.propTypes = {
  ...Button.propTypes
}
DropdownButton.displayName = 'Dropdown.Button'

DropdownCaret.defaultProps = {theme}
DropdownCaret.propTypes = {
  ...COMMON.propTypes,
  ...sx.propTypes
}
DropdownCaret.displayName = 'Dropdown.Caret'

Dropdown.defaultProps = {theme, ...Details.defaultProps}
Dropdown.propTypes = {
  ...Details.propTypes,
  ...COMMON.propTypes
}

export type DropdownCaretProps = ComponentProps<typeof DropdownCaret>
export type DropdownMenuProps = ComponentProps<typeof DropdownMenu>
export type DropdownItemProps = ComponentProps<typeof DropdownItem>
export default Object.assign(Dropdown, {
  Caret: DropdownCaret,
  Menu: DropdownMenu,
  Item: DropdownItem,
  Button: DropdownButton
})
