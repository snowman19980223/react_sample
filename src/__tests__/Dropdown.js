import React from 'react'
import Dropdown from '../Dropdown'
import {render} from '../utils/testing'
import {COMMON} from '../constants'

describe('Dropdown', () => {
  it('matches the snapshots', () => {
    expect(render(<Dropdown>hi</Dropdown>)).toMatchSnapshot()
    expect(render(<Dropdown title="hi">hello!</Dropdown>)).toMatchSnapshot()
  })

  it('implements system props', () => {
    expect(Dropdown).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(Dropdown).toSetDefaultTheme()
  })
})

describe('Dropdown.Item', () => {
  it('matches the snapshots', () => {
    expect(render(<Dropdown.Item>hi</Dropdown.Item>)).toMatchSnapshot()
    expect(render(<Dropdown.Item>hello!</Dropdown.Item>)).toMatchSnapshot()
  })

  it('implements system props', () => {
    expect(Dropdown.Item).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(Dropdown.Item).toSetDefaultTheme()
  })
})

describe('Dropdown.Menu', () => {
  it('matches the snapshots', () => {
    expect(
      render(
        <Dropdown.Menu>
          <li key="a">1</li>
          <li key="b">2</li>
          <li key="c">3</li>
        </Dropdown.Menu>
      )
    ).toMatchSnapshot()
    expect(
      render(
        <Dropdown.Menu>
          <li key={1}>1</li>
          <li key={2}>2</li>
          <li key={3}>3</li>
        </Dropdown.Menu>
      )
    ).toMatchSnapshot()
  })

  it('implements system props', () => {
    expect(Dropdown.Menu).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(Dropdown.Menu).toSetDefaultTheme()
  })
})
