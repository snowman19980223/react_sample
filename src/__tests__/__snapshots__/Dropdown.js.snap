// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches the snapshots 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  blacklist={
    Array [
      "css",
      "color",
      "bg",
      "m",
      "mt",
      "mr",
      "mb",
      "ml",
      "mx",
      "my",
      "p",
      "pt",
      "pr",
      "pb",
      "pl",
      "px",
      "py",
    ]
  }
  className="css-0 BtnGroup"
>
  <div
    className="BtnGroup-form c0"
    display="flex"
    is={[Function]}
  />
</div>
`;

exports[`Dropdown matches the snapshots 2`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  blacklist={
    Array [
      "css",
      "color",
      "bg",
      "m",
      "mt",
      "mr",
      "mb",
      "ml",
      "mx",
      "my",
      "p",
      "pt",
      "pr",
      "pb",
      "pl",
      "px",
      "py",
    ]
  }
  className="css-0 BtnGroup"
>
  <div
    className="BtnGroup-form c0"
    display="flex"
    is={[Function]}
  />
</div>
`;
